overlord/snapstate/autorefresh_gating.go

```diff
@@ -169,49 +197,60 @@ func HoldRefresh(st *state.State, gatingSnap string, holdDuration time.Duration,
 			}
 		}
 
-		lastRefreshTime, err := lastRefreshed(st, heldSnap)
-		if err != nil {
-			return 0, err
-		}
-
-		mp := maxPostponement - maxPostponementBuffer
-		maxDur := maxAllowedPostponement(gatingSnap, heldSnap, mp)
-
-		// calculate max hold duration that's left considering previous hold
-		// requests of this snap and last refresh time.
-		left := holdDurationLeft(now, lastRefreshTime, hold.FirstHeld, maxDur, mp)
-		if left <= 0 {
-			herr.SnapsInError[heldSnap] = HoldDurationError{
-				Err: fmt.Errorf("snap %q cannot hold snap %q anymore, maximum refresh postponement exceeded", gatingSnap, heldSnap),
+		if gatingSnap == "system" {
+			if holdDuration == 0 {
+				holdDuration = maxDuration
```

Comment:
This looks good, but it leads to a question: why can't we use `0` as a placeholder for "infinite" in the new code, too? (in HoldRefreshes(), for example)

Reply:
Holding forever needs to be internally consistent with `snap set system refresh.hold=forever` and setting a config takes a string. Since the gating code takes a duration there will always be some conversion, it's just a question of how we convert and where. We can take a zero time.Duration in HoldRefreshes but then we'd either have to do the conversion in the daemon (which seems wrong to me since the internal representation of "forever" should be an implementation detail of this package) or even further up the stack (but then the daemon would have to convert the zero duration to forever to pass into the configcore code)

Reply:
I see. Would it make sense to pass around a type like this?
```
type HoldDuration string
func (*HoldDuration) DurationFromNow() time.Duration
func (*HoldDuration) IsForever() bool
```
But if you think it's an unnecessary complication, forget it :-)

Reply:
That's an interesting idea but when I tried something like it, it required even more changes in the gating code. It was a new type that wrapped around a time or duration and had a "forever" flag (similar to what you wrote), which would replace hacky special values or limits for the forever case. But then we'd have to check that in a lot of places and we'd just be moving the special case checks from one place to the other =/ for practical purposes, mapping "forever" to 300 years is good enough

Reply:
Fair enough! :-)